crates/accelerate/src/convert_2q_block_matrix.rs

```diff
@@ -0,0 +1,72 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2022
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use pyo3::prelude::*;
+use pyo3::wrap_pyfunction;
+use pyo3::Python;
+
+use num_complex::Complex64;
+use numpy::ndarray::linalg::kron;
+use numpy::ndarray::{Array, Array2, ArrayView2};
+use numpy::{IntoPyArray, PyArray2, PyReadonlyArray2};
+
+/// Return the matrix Operator resulting from a block of Instructions.
+#[pyfunction]
+#[pyo3(text_signature = "(op_list, /")]
+pub fn blocks_to_matrix(
+    py: Python,
+    op_list: Vec<(PyReadonlyArray2<Complex64>, Vec<usize>)>,
+) -> PyResult<Py<PyArray2<Complex64>>> {
+    let identity: Array2<Complex64> = Array::eye(2);
+    let input_matrix = op_list[0].0.as_array();
+    let mut matrix: Array2<Complex64> = match op_list[0].1.as_slice() {
+        [0] => kron(&identity, &input_matrix),
+        [1] => kron(&input_matrix, &identity),
+        [0, 1] => input_matrix.to_owned(),
+        [1, 0] => change_basis(input_matrix),
+        _ => unreachable!(),
+    };
+    for (op_matrix, q_list) in op_list.into_iter().skip(1) {
+        let op_matrix = op_matrix.as_array();
+        let q_list = q_list.as_slice();
+        let result = match q_list {
+            [0] => Some(kron(&identity, &op_matrix)),
+            [1] => Some(kron(&op_matrix, &identity)),
+            [1, 0] => Some(change_basis(op_matrix)),
+            _ => None,
+        };
+        matrix = match result {
+            Some(result) => result.dot(&matrix),
+            None => op_matrix.dot(&matrix),
+        };
```

Comment:
Again, very minor, but if you're interested in style, this could equally be written as something like
```rust
let op_matrix = op_matrix.as_array();
let left = match q_list.as_slice() {
    [0] => kron(&identify, &op_matrix),
    [1] => kron(&op_matrix, &identity),
    [1, 0] => change_basis(op_matrix),
    [0, 1] => op_matrix,
    _ => unreachable!(),
};
matrix = left.dot(&matrix);
```

This is mostly just stylistic, though, so there's no need to do it if you don't want to / think your way is neater.

Reply:
We can get rid of the extra `q_list` variable. However, when it comes to that `[0,1]` case in the match function, the way the current approach is made without a `[0,1]` case is to be able to continue without necessarily copying the op_matrix (since it's an `ArrayView2` object, it would need to be returned as an `ArrayOwned2`.)

Just a funky rust way of avoiding copies here :v

Reply:
Ah, my mistake, I'd misread `op_matrix.as_array()` as being a clone operation (which tbh if it _had_ been, I should have commented on it!).